parsers/python.c

```diff
@@ -107,1237 +132,1052 @@ static kindOption PythonKinds[] = {
+								vStringPut (fq, '.');
+								vStringCat (fq, name->string);
+								makeSimplePythonRefTag (name, fq, K_UNKNOWN,
+								                        PYTHON_UNKNOWN_INDIRECTLY_IMPORTED,
+								                        XTAG_QUALIFIED_TAGS);
+								vStringDelete (fq);
+							}
+							/* Z */
+							makeSimplePythonTag (token, K_UNKNOWN);
+						}
+						else
+						{
+							/* import x as Y
+							 * ----------------------------
+							 * X = (kind:module, role:indirectly-imported)
+							 * Y = (kind:namespace)*/
+							/* X */
+							makeSimplePythonRefTag (name, NULL, K_MODULE,
+							                        PYTHON_MODULE_INDIRECTLY_IMPORTED,
+							                        XTAG_UNKNOWN);
+							/* Y */
+							makeSimplePythonTag (token, K_NAMESPACE);
+						}
+
+						copyToken (name, token);
+						readToken (token);
+					}
+				}
+				else
+				{
+					if (fromModule)
+					{
+						/* from x import Y
+						   --------------
+						   x = (kind:module,  role:namespace),
+						   Y = (kind:unknown, role:imported) */
+						/* Y */
+						makeSimplePythonRefTag (name, NULL, K_UNKNOWN,
+						                        PYTHON_MODULE_IMPORTED,
+						                        XTAG_UNKNOWN);
+						/* x.Y */
+						if (isXtagEnabled (XTAG_QUALIFIED_TAGS))
+						{
+							vString *fq = vStringNewCopy (fromModule->string);
+							vStringPut (fq, '.');
+							vStringCat (fq, name->string);
+							makeSimplePythonRefTag (name, fq, K_UNKNOWN,
+							                        PYTHON_MODULE_IMPORTED,
+							                        XTAG_QUALIFIED_TAGS);
+							vStringDelete (fq);
+						}
+					}
+					else
+					{
+						/* import X
+						   --------------
+						   X = (kind:module, role:imported) */
+						makeSimplePythonRefTag (name, NULL, K_MODULE,
+						                        PYTHON_MODULE_IMPORTED,
+						                        XTAG_UNKNOWN);
+					}
+				}
+
+				deleteToken (name);
 			}
 		}
-		if (!*ptr || *ptr == '=') return NULL;
-		if (*ptr == '(') {
-			return lastStart; /* if we stopped on a '(' we are done */
-		}
-		ptr = skipSpace(ptr);
-		lastStart = ptr;
-		while (*lastStart == '*') lastStart++;  /* cdef int *identifier */
+		while (token->type == ',');
+
+		if (parenthesized && token->type == ')')
+			readToken (token);
 	}
-	return NULL;
+
+	if (fromModule)
+		deleteToken (fromModule);
+
+	return FALSE;
 }
 
-/* checks if there is a lambda at position of cp, and return its argument list
- * if so.
- * We don't return the lambda name since it is useless for now since we already
- * know it when we call this function, and it would be a little slower. */
-static boolean varIsLambda (const char *cp, char **arglist)
+static boolean parseVariable (tokenInfo *const token)
 {
-	boolean is_lambda = FALSE;
+	/* In order to support proper tag type for lambdas in multiple
+	 * assignations, we first collect all the names, and then try and map
+	 * an assignation to it */
+	tokenInfo *nameTokens[8] = { NULL };
```

Comment:
8 name tokens is an arbitrary limit I made up, but I believe it's way large enough for non-crazy code.

However, we can easily augment this by simply changing this value here as the rest of the code only assumes it's at least 1, and otherwise relies on the actual element count.